module/zfs/vdev.c

```diff
@@ -3247,32 +3629,76 @@ vdev_space_update(vdev_t *vd, int64_t alloc_delta, int64_t defer_delta,
 	 * because the root vdev's psize-to-asize is simply the max of its
 	 * childrens', thus not accurate enough for us.
 	 */
-	ASSERT((dspace_delta & (SPA_MINBLOCKSIZE-1)) == 0);
-	ASSERT(vd->vdev_deflate_ratio != 0 || vd->vdev_isl2cache);
-	dspace_delta = (dspace_delta >> SPA_MINBLOCKSHIFT) *
-	    vd->vdev_deflate_ratio;
+	dspace_delta = vdev_deflated_space(vd, space_delta);
 
 	mutex_enter(&vd->vdev_stat_lock);
 	vd->vdev_stat.vs_alloc += alloc_delta;
 	vd->vdev_stat.vs_space += space_delta;
 	vd->vdev_stat.vs_dspace += dspace_delta;
+	if (special)
+		vd->vdev_stat.vs_special_alloc += alloc_delta;
 	mutex_exit(&vd->vdev_stat_lock);
 
-	if (mc == spa_normal_class(spa)) {
+	/* every class but log contributes to root space stats */
+	if (vd->vdev_mg != NULL && !vd->vdev_islog) {
 		mutex_enter(&rvd->vdev_stat_lock);
 		rvd->vdev_stat.vs_alloc += alloc_delta;
 		rvd->vdev_stat.vs_space += space_delta;
 		rvd->vdev_stat.vs_dspace += dspace_delta;
+		if (special)
+			rvd->vdev_stat.vs_special_alloc += alloc_delta;
 		mutex_exit(&rvd->vdev_stat_lock);
 	}
+	/* Note: metaslab_class_space_update moved to metaslab_space_update */
+}
+
+/*
+ * propagate metaslab block category changes to vdev state
+ * called from: metaslab_init, metaslab_fini, and metaslab_sync
+ */
+void
+vdev_category_space_update(vdev_t *vd, int64_t normal_assigned_delta,
+    int64_t special_assigned_delta)
+{
+	vdev_t *rvd = vd->vdev_spa->spa_root_vdev;
+
+	ASSERT(vd == vd->vdev_top);
 
-	if (mc != NULL) {
-		ASSERT(rvd == vd->vdev_parent);
-		ASSERT(vd->vdev_ms_count != 0);
+	mutex_enter(&vd->vdev_stat_lock);
+	vd->vdev_stat.vs_normal_assigned += normal_assigned_delta;
+	vd->vdev_stat.vs_special_assigned += special_assigned_delta;
+	mutex_exit(&vd->vdev_stat_lock);
+
+	mutex_enter(&rvd->vdev_stat_lock);
+	rvd->vdev_stat.vs_normal_assigned += normal_assigned_delta;
+	rvd->vdev_stat.vs_special_assigned += special_assigned_delta;
+	mutex_exit(&rvd->vdev_stat_lock);
+}
+
+#define	VDEV_MS_METADATA_RESERVE	10
 
-		metaslab_class_space_update(mc,
-		    alloc_delta, defer_delta, space_delta, dspace_delta);
+boolean_t
+vdev_category_space_full(spa_t *spa, metaslab_block_category_t category,
+    uint64_t request)
+{
+	metaslab_class_t *mc;
+	uint64_t space = 0, alloc = 0;
+
+	if (category == MS_CATEGORY_SMALL) {
+		mc = spa_special_class(spa);
+		space = metaslab_class_get_space(mc);
+		/* Maintain a reserve for metadata */
+		space = (space * (100 - VDEV_MS_METADATA_RESERVE)) / 100;
+		alloc = metaslab_class_get_alloc(mc) + request;
 	}
+
+	if (category == MS_CATEGORY_REGULAR) {
```

Comment:
nit: this could be an else if and alloc could be set outside the IFs after mc is set